### PR TITLE
Level control with setpoint and min/max bounds

### PR DESCRIPTION
**Pull request by SouthEndMusic**
Fixes https://github.com/Deltares/Ribasim/issues/447, fixes https://github.com/Deltares/Ribasim/issues/454.

To demonstrate the result of this PR we consider this model:

setpoint_with_bounds_example_model

The basin has a net forcing coming from the level boundary and infiltration. The basin has a setpoint and minimum and maximum bounds for its level. We want to activate one of the pumps when the level gets outside these bounds and bring it back to the setpoint. This has indeed been achieved:

setpoint_with_bounds_example

This has been implemented using crossing aware truth states (https://github.com/Deltares/Ribasim/issues/447), and the 'deceptive crossing' problem (https://github.com/Deltares/Ribasim/issues/454) has been solved by reformulating the `du` of the level condition basin after the parameter change. This is now done in a very inefficient way, but this can be improved (https://github.com/Deltares/Ribasim/issues/487).

Some remaining points:
- The setpoint in this case is implemented as a level condition, not as a `target_level` in `Basin \ static` which is only used for PID control. This might be confusing for users;
- Since this is a much more realistic example of control, I think this would be much more informative as the control example in `examples.ipynb`. The only possible drawback is that I constructed the net forcing in a bit of a hacky way, by starting with the forcing I wanted to have and distributing the positive and negative values over the flow boundary and precipitation respectively;
- I haven't tested this yet, I'm not sure what the best way would be to test this.

### CODE REVIEW

File: python/ribasim_testmodels/ribasim_testmodels/time.py [hunk 1]
```diff
@@ -102,3 +102,157 @@ def flow_boundary_time_model():
...
+        flow_boundary=flow_boundary,
+        discrete_control=discrete_control,
+        starttime="2020-01-01 00:00:00",
+        endtime="2021-01-01 00:00:00",
+    )
+
+    return model
+
+
+if __name__ == "__main__":
```
Comment (Huite) [inline python/ribasim_testmodels/ribasim_testmodels/time.py:253]: You probably want to remove this part.

File: python/ribasim_testmodels/ribasim_testmodels/time.py [hunk 2]
```diff
@@ -102,3 +102,157 @@ def flow_boundary_time_model():
     )
 
     return model
+
+
+def conditions_on_discrete_flow_model():
```
Comment (Huite) [inline python/ribasim_testmodels/ribasim_testmodels/time.py:107]: Could you add a docstring on what this model exactly represents (and tests)?

(As far as I can tell, it always listens to the flow of node 1, checks the flow, and as different values are exceeded, the pump is given a higher rate?)

Comment (SouthEndMusic) [reply python/ribasim_testmodels/ribasim_testmodels/time.py:107]: I'll ad a docstring (and give it a better name). What the model does is explained in https://github.com/Deltares/Ribasim/issues/447, and demonstrated above. We discussed this at length a few weeks back, about going back to the setpoint once the level is outside an accepted range.

**Comment by Huite** [comment]
I don't think I quite comprehend the changes, and so I think it's best to discuss this with @visr and you next week.

This adds additional truth states of "U" and "D". I'm afraid this is too hard a sell for a lot of users. 

Regarding your remarks:

> The setpoint in this case is implemented as a level condition, not as a target_level in Basin \ static which is only used for PID control. This might be confusing for users;

This is indeed unfortunate, and maybe it suggests that we should refactor. If different control mechanisms require different shapes of input, they just require separate inputs. I.e. the target level should be a property of the PID controller, not of the basin. 

> Since this is a much more realistic example of control, I think this would be much more informative as the control example in examples.ipynb. The only possible drawback is that I constructed the net forcing in a bit of a hacky way, by starting with the forcing I wanted to have and distributing the positive and negative values over the flow boundary and precipitation respectively;

We can make the example (the forcing) a lot simpler by including an additional node, e.g. a TabulatedRatingCurve, which causes the basin to slowly empty instead of the flow boundary. Then we would like to see the next phases:

1. The initial basin water level is too high. The pump is turned on. At the same time, the rating curve generates a trickle of water out of the basin (but is dwarfed by the pump).
2. The water basin level reaches the setpoint. The pump is turned off. Now the water level starts decreasing slowly due to the rating curve trickle.
3. The water basin level drops below the minimum value. The other pump is turned on.
4. The water basin level rises again and reaches the setpoint. The other pump is turned off.
5. The water basin level drop again.

This doesn't require any time dependent input, and the output level is much easier to interpret than the effect of the sinusoidal forcing.

> I haven't tested this yet, I'm not sure what the best way would be to test this.

I think the essential feature is that the model aims for a setpoint, but doesn't activate until another boundary value is crossed.
So if you use constant precipitation or constant evaporation, you expect to see a triangle wave. We can easily generate the expected values of the triangle wave we expect to see.

File: python/ribasim/ribasim/geometry/edge.py [hunk 3]
```diff
@@ -65,7 +65,9 @@ def write(self, directory: FilePath, modelname: str) -> None:
             gdf = gdf.set_geometry("geometry")
         else:
             gdf["geometry"] = None
-        gdf.to_file(directory / f"{modelname}.gpkg", layer=name)
+        gdf.to_file(
+            directory / f"{modelname}.gpkg", layer=name, driver="GPKG"
+        )  # , engine="pyogrio"
```
Comment (visr) [inline python/ribasim/ribasim/geometry/edge.py:70]:
```suggestion
        )
```

File: python/ribasim/ribasim/geometry/node.py [hunk 4]
```diff
@@ -63,7 +63,11 @@ def write(self, directory: FilePath, modelname: str) -> None:
         gdf = gpd.GeoDataFrame(data=dataframe)
         gdf = gdf.set_geometry("geometry")
 
-        gdf.to_file(directory / f"{modelname}.gpkg", layer=name)
+        gdf.to_file(
+            directory / f"{modelname}.gpkg",
+            layer=name,
+            driver="GPKG",  # engine="pyogrio"
```
Comment (visr) [inline python/ribasim/ribasim/geometry/node.py:69]:
```suggestion
            driver="GPKG",
```

**Review by visr — APPROVED** [review]
Looks good to me. Two minor suggestions.
